File: 2/js/src/MinHeap.ts
class MinHeap {
	heap: number[]

	constructor(dataArr ?: number[]) {
		if(dataArr != undefined) {
			this.heap = dataArr;
			this.buildHeap();
		} else {
			this.heap = [];
		}
	}

	swap(i: number, j: number) {
		const temp = this.heap[i];
		this.heap[i] = this.heap[j];
		this.heap[j] = temp;
	}

	getMin(i: number, j: number) {
		if(this.heap[i] < this.heap[j]) {
			return i;
		}
		return j;
	}

	buildHeap() {
		for(let i=Math.floor(this.heap.length/2 - 1); i>=0; i--) {
			this.siftDown(i);
		}
	}

	siftDown(i: number) {
		while(2*i + 1 < this.heap.length) {
			const left = 2*i + 1;
			const right = 2*i + 2;
			let min = left;

			if(right < this.heap.length) {
				min = this.getMin(left, right);
			}

			min = this.getMin(i, min);
			if(min == i) {
				return;
			}

			this.swap(min, i);
			i = min;
		}
	}

	siftUp(i: number) {
		while(i > 0) {
			const parent = Math.floor((i-1)/2);
			const min = this.getMin(parent, i);

			if(min == parent) {
				return;
			}

			this.swap(i, parent);
			i = parent;
		}
	}

	insert(data: number) {
		this.heap.push(data);
		this.siftUp(this.heap.length-1);
	}

	remove(): number {
		this.swap(0, this.heap.length-1);

		const deleted = this.heap.pop();
		if(deleted == undefined) {
			return -1;
		}

		this.siftDown(0);
		return deleted;
	}
}

export {MinHeap};

File: 2/java/MinHeap.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

public class MinHeap {
	private List<Integer> heap;

	MinHeap() {
		this.heap = new ArrayList<>();
	}	
	
	public void buildHeap(List<Integer> arr) {
		this.heap = arr;
		
		for(int i=this.heap.size()/2-1; i>=0; i--) {
			siftDown(i);
		}

	}	

	public void insert(int data) {
		this.heap.add(data);
		this.siftUp(heap.size()-1);
	}
	
	public int delete() {
		int deleted = heap.get(0);
		
		heap.set(0, heap.get(heap.size()-1));
		heap.remove(heap.size()-1);
		siftDown(0);

		return deleted;
	}
	
	private void siftUp(int index) {
		if(index == 0) {
			return;
		}

		int parent = (int)Math.floor((index-1)/2);

		while(heap.get(parent) > heap.get(index)) {
			int temp = heap.get(parent);
			heap.set(parent, heap.get(index));
			heap.set(index, temp);
			
			index = parent;
			parent = (int)Math.floor((index-1)/2);
		}
	}
	
	private void siftDown(int index) {
		while(index*2+1 < heap.size()) {

			int leftChildIndex = index*2+1;
			int rightChildIndex = index*2+2;
			int minChildIndex = getMinChildIndex(leftChildIndex, rightChildIndex); 

			int temp = heap.get(index);	
			
			if(temp > heap.get(minChildIndex)) {
				heap.set(index, heap.get(minChildIndex));	
				heap.set(minChildIndex, temp);
				index = minChildIndex;
			} else {
					break;
			}
			
		}
	}
	
	private int getMinChildIndex(int index1, int index2) {
		if(index2 == this.heap.size()) {
			return index1;
		}

		if(this.heap.get(index1) < this.heap.get(index2)) {
			return index1;
		} else {
				return index2;
		}
	}

	public void print() {
		for(int i : this.heap) {
			System.out.print(i + " ");
		}
		System.out.println();
	}

	public static void main(String args[]) {
		MinHeap mh = new MinHeap(); 
		List<Integer> arr = new ArrayList<Integer>(Arrays.asList(new Integer[]{30,102,23,17,18,9,44,12,31}));
		
		mh.buildHeap(arr);
		mh.print();
		
		/*
			 int arr[] = {8,12,23,17,31,30,44,102,18};

			 for(int i : arr) {
			 mh.insert(i);
			 }
			 mh.insert(9);
			 mh.delete();
			 mh.print();
	*/
	}
}
